helios-client/src/main/java/com/spotify/helios/client/DefaultRequestDispatcher.java

```diff
@@ -169,107 +161,67 @@ private HttpURLConnection connect(final URI uri, final String method, final byte
                                     final Map<String, List<String>> headers)
       throws URISyntaxException, IOException, TimeoutException, InterruptedException,
              HeliosException {
-    final long deadline = currentTimeMillis() + RETRY_TIMEOUT_MILLIS;
-    final int offset = ThreadLocalRandom.current().nextInt();
 
-    while (currentTimeMillis() < deadline) {
-      final List<URI> endpoints = endpointSupplier.get();
-      if (endpoints.isEmpty()) {
-        throw new RuntimeException("failed to resolve master");
-      }
-      log.debug("endpoint uris are {}", endpoints);
-
-      // Resolve hostname into IPs so client will round-robin and retry for multiple A records.
-      // Keep a mapping of IPs to hostnames for TLS verification.
-      final List<URI> ipEndpoints = Lists.newArrayList();
-      final Map<URI, URI> ipToHostnameUris = Maps.newHashMap();
-
-      for (final URI hnUri : endpoints) {
-        try {
-          final InetAddress[] ips = InetAddress.getAllByName(hnUri.getHost());
-          for (final InetAddress ip : ips) {
-            final URI ipUri = new URI(
-                hnUri.getScheme(), hnUri.getUserInfo(), ip.getHostAddress(), hnUri.getPort(),
-                hnUri.getPath(), hnUri.getQuery(), hnUri.getFragment());
-            ipEndpoints.add(ipUri);
-            ipToHostnameUris.put(ipUri, hnUri);
+      final Endpoint endpoint = endpointIterator.next();
+      final URI endpointUri = endpoint.getUri();
+      final String fullpath = endpointUri.getPath() + uri.getPath();
+
+      final String uriScheme = endpointUri.getScheme();
+
+      final URI ipUri = new URI(
+          uriScheme, endpointUri.getUserInfo(), endpoint.getIp().getHostAddress(),
+          endpointUri.getPort(), fullpath, uri.getQuery(), null);
+
+      AgentProxy agentProxy = null;
+      Deque<Identity> identities = Queues.newArrayDeque();
```

Comment:
I think this could just be a regular Queue / LinkedList, we only insert at the end and remove from the head